### PR TITLE
OCPBUGS-33082: Add new use cases that requires anonymization

### PR DESCRIPTION
**Pull request by ncaak**
This PR implements a new data enhancement to...

## Categories

- [X] Bugfix
- [ ] Data Enhancement
- [ ] Feature
- [ ] Backporting
- [ ] Others (CI, Infrastructure, Documentation)

## Sample Archive

- ``

## Documentation

- `docs/gathered-data.md`

## Unit Tests

- `pkg/gatherers/clusterconfig/gather_cluster_infrastructure_test.go`

## Privacy

Yes. There are no sensitive data in the newly collected information.

## Changelog

- Yes

## Breaking Changes

- No

## References

https://issues.redhat.com/browse/OCPBUGS-33082

### CODE REVIEW

File: pkg/anonymization/anonymizer.go [hunk 1]
```diff
@@ -362,17 +343,19 @@ func (anonymizer *Anonymizer) AnonymizeMemoryRecord(memoryRecord *record.MemoryR
...
-			memoryRecord.Data,
-			[]byte(anonymizer.clusterBaseDomain),
-			[]byte(ClusterBaseDomainPlaceholder),
-		)
-		memoryRecord.Name = strings.ReplaceAll(
-			memoryRecord.Name,
-			anonymizer.clusterBaseDomain,
-			ClusterBaseDomainPlaceholder,
-		)
+	if len(anonymizer.sensitiveValues) != 0 {
```
Comment (tremes) [inline pkg/anonymization/anonymizer.go:346]: I think this condition is not needed.

File: pkg/anonymization/builder.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	k8snet "k8s.io/utils/net"
+)
+
+type AnonBuilder struct {
+	anon     Anonymizer
+	networks []string
+}
+
+func (b *AnonBuilder) AddSensitiveValue(value, placeholder string) *AnonBuilder {
```
Comment (tremes) [inline pkg/anonymization/builder.go:21]: I think at least this function would deserve some docs/comment.

File: pkg/anonymization/builder.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	if b.anon.sensitiveValues == nil {
+		b.anon.sensitiveValues = make(map[string]string)
+	}
+	v := strings.TrimSpace(value)
+	if value != "" {
+		b.anon.sensitiveValues[v] = placeholder
+	}
```
Comment (tremes) [inline pkg/anonymization/builder.go:28]: Perhaps we could do at the beginning of the function:
```
if value == "" {
  return b
}
```

File: pkg/anonymization/builder_test.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// When
+			test, err := tc.builder.Build()
+
+			// Assert
+			assert.NoError(t, err)
+			assert.IsType(t, Anonymizer{}, *test)
+			if tc.sensitiveValues != nil {
```
Comment (tremes) [inline pkg/anonymization/builder_test.go:100]: Are these conditions required? I guess when you provide `nil` input, you can also expect `nil` output or?

File: pkg/utils/apiserver.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+//
+// It uses this API https://docs.openshift.com/container-platform/4.7/rest_api/config_apis/dns-config-openshift-io-v1.html
+func GetClusterAPIServerInfo(ctx context.Context, configClient configv1client.ConfigV1Interface) ([]string, error) {
+	infra, err := configClient.Infrastructures().Get(ctx, "cluster", metav1.GetOptions{})
+	if err != nil {
+		return []string{}, err
+	}
+
+	urls := []string{}
+	for _, url := range []string{infra.Status.APIServerURL, infra.Status.APIServerInternalURL} {
```
Comment (tremes) [inline pkg/utils/apiserver.go:24]: This will cause panic if the state is not set (probably an edge case).

File: pkg/anonymization/builder_test.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+			if tc.kubeClient != nil {
+				assert.Equal(t, tc.kubeClient, test.gatherKubeClient)
+			}
+			if tc.networkClient != nil {
+				assert.Equal(t, tc.networkClient, test.networkClient)
+			}
+			if len(tc.networks) > 0 {
+				assert.Equal(t, len(tc.networks), len(test.networks))
+			}
+			if tc.runningInCluster {
```
Comment (tremes) [inline pkg/anonymization/builder_test.go:121]: I think only this condition is really required :)

File: pkg/anonymization/builder.go (hunk 2 above)
Comment (tremes) [reply pkg/anonymization/builder.go:21]: What about to call it `WithSensitiveValue` (to keep the "pattern" below) ?

File: pkg/anonymization/builder_test.go (hunk 6 above)
Comment (ncaak) [reply pkg/anonymization/builder_test.go:109]: Solved in bc66284efa66957e06efbc3efd2c1f893793085e, thanks for the review

File: pkg/anonymization/builder_test.go (hunk 4 above)
Comment (ncaak) [reply pkg/anonymization/builder_test.go:100]: Solved in bc66284efa66957e06efbc3efd2c1f893793085e, thanks for the review

File: pkg/anonymization/builder.go (hunk 3 above)
Comment (ncaak) [reply pkg/anonymization/builder.go:27]: Solved in bc66284efa66957e06efbc3efd2c1f893793085e, thanks for the review

File: pkg/anonymization/builder.go (hunk 2 above)
Comment (ncaak) [reply pkg/anonymization/builder.go]: Solved in bc66284efa66957e06efbc3efd2c1f893793085e, thanks for the review

File: pkg/anonymization/anonymizer.go (hunk 1 above)
Comment (ncaak) [reply pkg/anonymization/anonymizer.go]: Solved in bc66284efa66957e06efbc3efd2c1f893793085e, thanks for the review

File: pkg/utils/apiserver.go (hunk 5 above)
Comment (ncaak) [reply pkg/utils/apiserver.go:24]: I'm going to add some unit tests there and refactor the logic. Thanks for the review.

Comment (ncaak) [reply pkg/utils/apiserver.go:28]: Solved in 1e7431666a292f50e34e76fa700279364f700078

File: pkg/utils/apiserver.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+// For example, given the base domain `openshift.example.com`,
+// an API server could be in `api.hypershift.local`.
+//
+// It uses this API https://docs.openshift.com/container-platform/4.7/rest_api/config_apis/dns-config-openshift-io-v1.html
+func GetClusterAPIServerInfo(ctx context.Context, configClient configv1client.ConfigV1Interface) ([]string, error) {
+	infra, err := configClient.Infrastructures().Get(ctx, "cluster", metav1.GetOptions{})
+	if err != nil {
+		return []string{}, err
+	}
+	if infra == nil {
```
Comment (tremes) [inline pkg/utils/apiserver.go:23]: The idea below was to check `infra.Status == nil` to prevent possible panic.

Comment (ncaak) [reply pkg/utils/apiserver.go:23]: I know, but through the unit tests I couldn't set it as nil because of a golang failure. Do you think it's possible for the struct to be initialized as `nil`, even if golang causes an error, as to reach there that way?

Comment (tremes) [reply pkg/utils/apiserver.go:23]: Oh OK so the Status attribute is not a pointer so it can't be nil. Sorry for the noise.

Comment (ncaak) [reply pkg/utils/apiserver.go:23]: No problem. Actually, `infra` object could be nil, so I caught that possibility. Thanks for the revision.
